### ISSUE TITLE
Cascading selects do not work in our version of Enketo

### ISSUE THREAD

**Issue by sglangevin** [medic/medic-webapp#3624]
@alxndrsn's notes from the original issue (https://github.com/medic/medic-projects/issues/2276):

From enketo-xslt:

> WARNING: Itemset support is experimental. Make sure to test whether they do what you want.
–https://github.com/enketo/enketo-xslt/blob/master/xsl/openrosa2html5form.xsl#L186

I guess they don't do what we want.

It's seems this is due to an XPath 2.0 operations in the XSL stylesheet, so this probably works in enketo-core.

Expressions/functions that will need rewriting:
- `substring-after()`
- `dyn:evaluate()`

--------
I've tested this on Ona previewer, so it appears that cascading selects are working in enketo-core. Here is a sample form that works in Ona and can be used for testing:
```
<?xml version="1.0"?>
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:jr="http://openrosa.org/javarosa" xmlns:orx="http://openrosa.org/xforms" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <h:head>
    <h:title>Outreach</h:title>
    <model>
      <itext>
        <translation lang="default">
          <text id="static_instance-sites-22">
            <value>Goma</value>
          </text>
          <text id="static_instance-sites-23">
            <value>Other</value>
          </text>
          <text id="static_instance-sites-20">
            <value>Kabonga</value>
          </text>
          <text id="static_instance-sites-21">
            <value>Other</value>
          </text>
          <text id="static_instance-sites-9">
            <value>Lusolo</value>
          </text>
          <text id="static_instance-sites-8">
            <value>Kirongwe</value>
          </text>
          <text id="static_instance-sites-7">
            <value>Rusumo</value>
          </text>
          <text id="static_instance-sites-6">
            <value>Holili</value>
          </text>
          <text id="static_instance-sites-5">
            <value>Other</value>
          </text>
          <text id="static_instance-sites-4">
            <value>Muhurubay</value>
          </text>
          <text id="static_instance-sites-3">
            <value>Taveta</value>
          </text>
          <text id="static_instance-sites-2">
            <value>Sio Port</value>
          </text>
          <text id="static_instance-sites-1">
            <value>Malaba</value>
          </text>
          <text id="static_instance-sites-0">
            <value>Busia</value>
          </text>
          <text id="static_instance-sites-13">
            <value>Majanji</value>
          </text>
          <text id="static_instance-sites-12">
            <value>Malaba</value>
          </text>
          <text id="static_instance-sites-11">
            <value>Busia</value>
          </text>
          <text id="static_instance-sites-10">
            <value>Other</value>
          </text>
          <text id="static_instance-sites-17">
            <value>Gatuna</value>
          </text>
          <text id="static_instance-sites-16">
            <value>Rubavu</value>
          </text>
          <text id="static_instance-sites-15">
            <value>Other</value>
          </text>
          <text id="static_instance-sites-14">
            <value>Katuna</value>
          </text>
          <text id="static_instance-sites-19">
            <value>Other</value>
          </text>
          <text id="static_instance-sites-18">
            <value>Rusumo</value>
          </text>
          <text id="static_instance-countries-5">
            <value>DRC</value>
          </text>
          <text id="static_instance-countries-4">
            <value>Burundi</value>
          </text>
          <text id="static_instance-countries-6">
            <value>Other</value>
          </text>
          <text id="static_instance-countries-1">
            <value>Uganda</value>
          </text>
          <text id="static_instance-countries-0">
            <value>Kenya</value>
          </text>
          <text id="static_instance-countries-3">
            <value>Rwanda</value>
          </text>
          <text id="static_instance-countries-2">
            <value>Tanzania</value>
          </text>
        </translation>
      </itext>
      <instance>
        <outreach delimiter="#" id="outreach" prefix="J1!outreach!" version="6/9/2017">
          <country/>
          <country_other/>
          <site/>
          <site_other/>
          <meta>
            <instanceID/>
          </meta>
        </outreach>
      </instance>
      <instance id="sites">
        <root>
          <item>
            <itextId>static_instance-sites-0</itextId>
            <country>kenya</country>
            <name>busia</name>
          </item>
          <item>
            <itextId>static_instance-sites-1</itextId>
            <country>kenya</country>
            <name>malaba</name>
          </item>
          <item>
            <itextId>static_instance-sites-2</itextId>
            <country>kenya</country>
            <name>sio_port</name>
          </item>
          <item>
            <itextId>static_instance-sites-3</itextId>
            <country>kenya</country>
            <name>taveta</name>
          </item>
          <item>
            <itextId>static_instance-sites-4</itextId>
            <country>kenya</country>
            <name>muhurubay</name>
          </item>
          <item>
            <itextId>static_instance-sites-5</itextId>
            <country>kenya</country>
            <name>other</name>
          </item>
          <item>
            <itextId>static_instance-sites-6</itextId>
            <country>tanzania</country>
            <name>holili</name>
          </item>
          <item>
            <itextId>static_instance-sites-7</itextId>
            <country>tanzania</country>
            <name>rusumo</name>
          </item>
          <item>
            <itextId>static_instance-sites-8</itextId>
            <country>tanzania</country>
            <name>kirongwe</name>
          </item>
          <item>
            <itextId>static_instance-sites-9</itextId>
            <country>tanzania</country>
            <name>lusolo</name>
          </item>
          <item>
            <itextId>static_instance-sites-10</itextId>
            <country>tanzania</country>
            <name>other</name>
          </item>
          <item>
            <itextId>static_instance-sites-11</itextId>
            <country>uganda</country>
            <name>busia</name>
          </item>
          <item>
            <itextId>static_instance-sites-12</itextId>
            <country>uganda</country>
            <name>malaba</name>
          </item>
          <item>
            <itextId>static_instance-sites-13</itextId>
            <country>uganda</country>
            <name>majanji</name>
          </item>
          <item>
            <itextId>static_instance-sites-14</itextId>
            <country>uganda</country>
            <name>katuna</name>
          </item>
          <item>
            <itextId>static_instance-sites-15</itextId>
            <country>uganda</country>
            <name>other</name>
          </item>
          <item>
            <itextId>static_instance-sites-16</itextId>
            <country>rwanda</country>
            <name>rubavu</name>
          </item>
          <item>
            <itextId>static_instance-sites-17</itextId>
            <country>rwanda</country>
            <name>gatuna</name>
          </item>
          <item>
            <itextId>static_instance-sites-18</itextId>
            <country>rwanda</country>
            <name>rusumo</name>
          </item>
          <item>
            <itextId>static_instance-sites-19</itextId>
            <country>rwanda</country>
            <name>other</name>
          </item>
          <item>
            <itextId>static_instance-sites-20</itextId>
            <country>burundi</country>
            <name>kabonga</name>
          </item>
          <item>
            <itextId>static_instance-sites-21</itextId>
            <country>burundi</country>
            <name>other</name>
          </item>
          <item>
            <itextId>static_instance-sites-22</itextId>
            <country>drc</country>
            <name>goma</name>
          </item>
          <item>
            <itextId>static_instance-sites-23</itextId>
            <country>drc</country>
            <name>other</name>
          </item>
        </root>
      </instance>
      <instance id="countries">
        <root>
          <item>
            <itextId>static_instance-countries-0</itextId>
            <name>kenya</name>
          </item>
          <item>
            <itextId>static_instance-countries-1</itextId>
            <name>uganda</name>
          </item>
          <item>
            <itextId>static_instance-countries-2</itextId>
            <name>tanzania</name>
          </item>
          <item>
            <itextId>static_instance-countries-3</itextId>
            <name>rwanda</name>
          </item>
          <item>
            <itextId>static_instance-countries-4</itextId>
            <name>burundi</name>
          </item>
          <item>
            <itextId>static_instance-countries-5</itextId>
            <name>drc</name>
          </item>
          <item>
            <itextId>static_instance-countries-6</itextId>
            <name>other</name>
          </item>
        </root>
      </instance>
      <bind nodeset="/outreach/country" required="true()" type="select1"/>
      <bind nodeset="/outreach/country_other" relevant="selected( /outreach/country , 'other')" type="string"/>
      <bind nodeset="/outreach/site" required="true()" type="select1"/>
      <bind nodeset="/outreach/site_other" relevant="selected( /outreach/site , 'other')" type="string"/>
      <bind calculate="concat('uuid:', uuid())" nodeset="/outreach/meta/instanceID" readonly="true()" type="string"/>
    </model>
  </h:head>
  <h:body class="pages">
    <select1 appearance="minimal" ref="/outreach/country">
      <label>Country</label>
      <item>
        <label>Kenya</label>
        <value>kenya</value>
      </item>
      <item>
        <label>Uganda</label>
        <value>uganda</value>
      </item>
      <item>
        <label>Tanzania</label>
        <value>tanzania</value>
      </item>
      <item>
        <label>Rwanda</label>
        <value>rwanda</value>
      </item>
      <item>
        <label>Burundi</label>
        <value>burundi</value>
      </item>
      <item>
        <label>DRC</label>
        <value>drc</value>
      </item>
      <item>
        <label>Other</label>
        <value>other</value>
      </item>
    </select1>
    <input ref="/outreach/country_other">
      <label>Specify other</label>
    </input>
    <select1 appearance="minimal" ref="/outreach/site">
      <label>Site</label>
      <itemset nodeset="instance('sites')/root/item[country =  /outreach/country ]">
        <value ref="name"/>
        <label ref="jr:itext(itextId)"/>
      </itemset>
    </select1>
    <input ref="/outreach/site_other">
      <label>Specify other</label>
    </input>
  </h:body>
</h:html>
```

**Comment by sglangevin** [comment]
@alxndrsn this one has been sitting for a while and I've received requests from 2-3 projects for cascading selects in our app. Will you be able to get to this one sometime this week?

**Comment by alxndrsn** [comment]
@sglangevin yes, it's top of the pile 🙂

**Comment by alxndrsn** [comment]
As far as I can tell, `substring-after()` works fine in latest FF and Chrome, and in Chrome 30:

<img width="866" alt="screen shot 2017-10-03 at 11 34 58" src="https://user-images.githubusercontent.com/191496/31118802-fa30d544-a82e-11e7-96e7-f53cb310391e.png">

**Comment by alxndrsn** [comment]
As far as I can tell, `substring-after()` works fine in latest FF and Chrome, and in Chrome 30:

<img width="866" alt="screen shot 2017-10-03 at 11 34 58" src="https://user-images.githubusercontent.com/191496/31118802-fa30d544-a82e-11e7-96e7-f53cb310391e.png">

I think the problem at that point was actually with `str:replace`, which we _should_ be able to replace with the template `str-replace` (although it didn't seem to work immediately for me).

However, the bigger problem is `dyn:evaluate()` - I'm not sure how we can re-implement this without direct access to the `XSLProcessor`.  I'll read up on the Chrome docs to see if there's a way we can add this function.

In the meantime, the following structure should work in place of the example form.  Would this approach be sufficient for the time being?

```
<?xml version="1.0"?>
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:jr="http://openrosa.org/javarosa" xmlns:orx="http://openrosa.org/xforms" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <h:head>
    <h:title>Test Form - Cascading Selects</h:title>
    <model>
      <instance>
        <outreach delimiter="#" id="test-cascading-selects" prefix="J1!outreach!" version="6/9/2017">
          <country/>
          <country_other/>
          <site>
            <kenya/>
            <uganda/>
            <other/>
          </site>
          <selected_site/>
          <meta>
            <instanceID/>
          </meta>
        </outreach>
      </instance>
      <bind nodeset="/outreach/country" required="true()" type="select1"/>
      <bind nodeset="/outreach/country_other" relevant="selected( /outreach/country , 'other')" type="string"/>
      <bind nodeset="/outreach/site/kenya" relevant="selected( /outreach/country , 'kenya')" required="true()" type="select1"/>
      <bind nodeset="/outreach/site/uganda" relevant="selected( /outreach/country , 'uganda')" required="true()" type="select1"/>
      <bind nodeset="/outreach/site/other" relevant="( /outreach/country = 'other' ) or ( /outreach/country = 'kenya' and selected( /outreach/site/kenya, 'other') ) or ( /outreach/country = 'uganda' and selected( /outreach/site/uganda, 'other') )" type="string"/>
      <bind nodeset="/outreach/selected_site" type="string" calculate=" if( /outreach/country = 'kenya', if( /outreach/site/kenya = 'other', /outreach/site/other, /outreach/site/kenya ), if( /outreach/country = 'uganda', if( /outreach/site/uganda = 'other', /outreach/site/other, /outreach/site/uganda ), /outreach/site/other ) )"/>
      <bind calculate="concat('uuid:', uuid())" nodeset="/outreach/meta/instanceID" readonly="true()" type="string"/>
    </model>
  </h:head>
  <h:body class="pages">
    <select1 appearance="minimal" ref="/outreach/country">
      <label>Country</label>
      <item>
        <label>Kenya</label>
        <value>kenya</value>
      </item>
      <item>
        <label>Uganda</label>
        <value>uganda</value>
      </item>
      <item>
        <label>Other</label>
        <value>other</value>
      </item>
    </select1>
    <input ref="/outreach/country_other">
      <label>Specify country</label>
    </input>
    <select1 appearance="minimal" ref="/outreach/site/kenya">
      <label>Site</label>
      <item>
        <label>Malaba</label>
        <value>malaba</value>
      </item>
      <item>
        <label>Sio Port</label>
        <value>sio_port</value>
      </item>
      <item>
        <label>Taveta</label>
        <value>taveta</value>
      </item>
      <item>
        <label>Muhurubay</label>
        <value>muhurubay</value>
      </item>
      <item>
        <label>Other</label>
        <value>other</value>
      </item>
    </select1>
    <select1 appearance="minimal" ref="/outreach/site/uganda">
      <label>Site</label>
      <item>
        <label>Malaba</label>
        <value>malaba</value>
      </item>
      <item>
        <label>Majanji</label>
        <value>majanji</value>
      </item>
      <item>
        <label>Katuna</label>
        <value>katuna</value>
      </item>
      <item>
        <label>Other</label>
        <value>other</value>
      </item>
    </select1>
    <input ref="/outreach/site/other">
      <label>Specify site</label>
    </input>
  </h:body>
</h:html>
```

**Comment by sglangevin** [comment]
I'll take a look at this tomorrow and see if I can produce that XML from an XLS form.

**Comment by alxndrsn** [comment]
@sglangevin 
Here's an alternative that will work, but at the cost of removing translation support from `<itemset>`.  I don't think this feature currently works anyway in webapp on Chrome, so it shouldn't affect existing projects.

```
<?xml version="1.0"?>
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:jr="http://openrosa.org/javarosa" xmlns:orx="http://openrosa.org/xforms" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <h:head>
    <h:title>Outreach</h:title>
    <model>
      <instance>
        <outreach delimiter="#" id="outreach" prefix="J1!outreach!" version="6/9/2017">
          <country/>
          <country_other/>
          <site/>
          <site_other/>
          <meta>
            <instanceID/>
          </meta>
        </outreach>
      </instance>
      <instance id="sites">
        <root>
          <item>
            <country>kenya</country>
            <label>busia</label>
            <value>Busia</value>
          </item>
          <item>
            <country>kenya</country>
            <label>malaba</label>
            <value>Malaba</value>
          </item>
          <item>
            <country>kenya</country>
            <label>sio_port</label>
            <value>Sio_port</value>
          </item>
          <item>
            <country>kenya</country>
            <label>taveta</label>
            <value>Taveta</value>
          </item>
          <item>
            <country>kenya</country>
            <label>muhurubay</label>
            <value>Muhurubay</value>
          </item>
          <item>
            <country>kenya</country>
            <label>other</label>
            <value>Other</value>
          </item>
          <item>
            <country>tanzania</country>
            <label>holili</label>
            <value>Holili</value>
          </item>
          <item>
            <country>tanzania</country>
            <label>rusumo</label>
            <value>Rusumo</value>
          </item>
          <item>
            <country>tanzania</country>
            <label>kirongwe</label>
            <value>Kirongwe</value>
          </item>
          <item>
            <country>tanzania</country>
            <label>lusolo</label>
            <value>Lusolo</value>
          </item>
          <item>
            <country>tanzania</country>
            <label>other</label>
            <value>Other</value>
          </item>
          <item>
            <country>uganda</country>
            <label>busia</label>
            <value>Busia</value>
          </item>
          <item>
            <country>uganda</country>
            <label>malaba</label>
            <value>Malaba</value>
          </item>
          <item>
            <country>uganda</country>
            <label>majanji</label>
            <value>Majanji</value>
          </item>
          <item>
            <country>uganda</country>
            <label>katuna</label>
            <value>Katuna</value>
          </item>
          <item>
            <country>uganda</country>
            <label>other</label>
            <value>Other</value>
          </item>
          <item>
            <country>rwanda</country>
            <label>rubavu</label>
            <value>Rubavu</value>
          </item>
          <item>
            <country>rwanda</country>
            <label>gatuna</label>
            <value>Gatuna</value>
          </item>
          <item>
            <country>rwanda</country>
            <label>rusumo</label>
            <value>Rusumo</value>
          </item>
          <item>
            <country>rwanda</country>
            <label>other</label>
            <value>Other</value>
          </item>
          <item>
            <country>burundi</country>
            <label>kabonga</label>
            <value>Kabonga</value>
          </item>
          <item>
            <country>burundi</country>
            <label>other</label>
            <value>Other</value>
          </item>
          <item>
            <country>drc</country>
            <label>goma</label>
            <value>Goma</value>
          </item>
          <item>
            <country>drc</country>
            <label>other</label>
            <value>Other</value>
          </item>
        </root>
      </instance>
      <bind nodeset="/outreach/country" required="true()" type="select1"/>
      <bind nodeset="/outreach/country_other" relevant="selected( /outreach/country , 'other')" type="string"/>
      <bind nodeset="/outreach/site" required="true()" type="select1"/>
      <bind nodeset="/outreach/site_other" relevant="selected( /outreach/site , 'other')" type="string"/>
      <bind calculate="concat('uuid:', uuid())" nodeset="/outreach/meta/instanceID" readonly="true()" type="string"/>
    </model>
  </h:head>
  <h:body class="pages">
    <select1 appearance="minimal" ref="/outreach/country">
      <label>Country</label>
      <item>
        <label>Kenya</label>
        <value>kenya</value>
      </item>
      <item>
        <label>Uganda</label>
        <value>uganda</value>
      </item>
      <item>
        <label>Tanzania</label>
        <value>tanzania</value>
      </item>
      <item>
        <label>Rwanda</label>
        <value>rwanda</value>
      </item>
      <item>
        <label>Burundi</label>
        <value>burundi</value>
      </item>
      <item>
        <label>DRC</label>
        <value>drc</value>
      </item>
      <item>
        <label>Other</label>
        <value>other</value>
      </item>
    </select1>
    <input ref="/outreach/country_other">
      <label>Specify other</label>
    </input>
    <select1 appearance="minimal" ref="/outreach/site">
      <label>Site</label>
      <itemset nodeset="instance('sites')/root/item[country =  /outreach/country ]">
        <label ref="label"/>
        <value ref="value"/>
      </itemset>
    </select1>
    <input ref="/outreach/site_other">
      <label>Specify other</label>
    </input>
  </h:body>
</h:html>
```

**Comment by sglangevin** [comment]
I think issue is that none of these can be produced from an xlsform and we need to be able to create something in xlsform that can be converted to the proper xml. When converting, I get the structure I pasted at the top of this thread, even if I'm not including any translations anywhere it still generates that in the xml. Is there something we can do in medic-configurer to help us along so that we don't have to manually edit the xml? That isn't workable for tech leads.

**Comment by alxndrsn** [comment]
I managed to eliminate the `dyn:evaluate()` call in the XSL file by assuming that XLSForm will always generate instances like this:
```
      <instance id="whatever">
        <root>
          <item>
```

**Comment by garethbowen** [comment]
@SCdF I've done my code review - leaving assigned to you for the other part.

**Comment by SCdF** [comment]
LGTM

**Comment by sglangevin** [comment]
Awesome, working well! Not sure if this was merged before 2.14.x was branched. If not, please backport. Moving to ready.

**Comment by alxndrsn** [comment]
Already backported 🙂
